### ISSUE TITLE
suspicious memory behavior in RgbdSensor body_pose_in_world_output_port()

### ISSUE THREAD

**Issue by RussTedrake** [RobotLocomotion/drake#15973]
The recent binding of `body_pose_in_world` output port (`Isometry3` => `RigidTransform`) might have some issues.  Or at least it led to mysterious memory-error looking issues on the pset students are working on this weekend.  https://deepnote.com/project/61-Label-Generation-yRBDseApSxaZJJ96ynWgEA/%2Flabel_generation.ipynb/#00003-44884cee-ca37-4962-87d8-267d6bec36b9

See the lines
```
        cam0_pose = cam0.GetOutputPort('X_WB').Eval(cam0_context)
        self.X_WC = RigidTransform(
            quaternion=Quaternion(cam0_pose[3:]), p=cam0_pose[0:3])
        # TODO(russt): Figure out why the replacement below led to memory errors
        #self.X_WC = cam0.GetOutputPort('body_pose_in_world').Eval(cam0_context)
```

The reported issue (which I've confirmed) is that if you print `X_WC` after that cell where it is initially defined, and then again a few cells later, and run the cells a few times, then you see that it's value changes in mysterious ways (e.g. it is no longer a valid RotationMatrix).

I'm sorry it's not a minimal repro yet.  @jwnimmer-tri -- might you be willing to take a quick look?

**Comment by RussTedrake** [comment]
That makes sense.  Thank you.  It is consistent with the fact that `Isometry3` version didn't have problems (it was doing the copy).  I agree that the contract I expected was the reference counting we do would have prevented it.  Should I leave this open so that we can track it down?

**Comment by jwnimmer-tri** [comment]
Ideally yes, assuming we can get it down into a reasonable reproduction scenario.

**Comment by RussTedrake** [comment]
I'm seeing lots of people running into something that looks like this issue.  They create some variable, don't realize it's a reference, and end up letting the containing diagram get garbage collected.  I'm really do think this is high priority... it's proving to be a real pain point.

**Comment by jwnimmer-tri** [comment]
To investigate this, we'll need a reproducible example that can be run locally standalone (even if not minimal yet), not as part of a notebook.  Ideally, it would only import from pydrake (and not manipulation), but we can work around that by forking a bunch of extra files if need be, and then cut it down to size once we have it up and running.

I tried to copy the notebook code and its helper functions into a standalone Python unit test file, but my attempt at that test case passes for me in both Release and Debug mode, when run as `bazel test` on Ubuntu Focal 20.04 locally.

Here is the WIP branch https://github.com/jwnimmer-tri/drake/tree/crash-test with that attempt.

Does anyone have a standalone, reproducible example of this failure?

**Comment by RussTedrake** [comment]
Sorry.. you diagnosed the issue, so I thought you knew how to repro.  My first minimal attempt reproduced the error, at least in a jupyter notebook.

In 1 cell, do
```
from pydrake.all import (
    ConstantValueSource, AbstractValue, RigidTransform
)

def myfun():
    system = ConstantValueSource(AbstractValue.Make(RigidTransform()))
    context = system.CreateDefaultContext()
    return system.get_output_port().Eval(context)

X = myfun()
```
then in a second cell, do
```
print(X)
```

I get outputs like
```
RigidTransform(
  R=RotationMatrix([
    [0.0, 1.500089289359168e+248, 1.0471813005185073e-142],
    [0.0, 4.5062008296066193e-144, 1.14428494433879e+243],
    [0.0, 4.824123284039802e+228, 1.3561721836670498e+248],
  ]),
  p=[2.78225510907157e+296, 2.2089474909638482e+161, 6.81019671049766e-310],
)
```
(which is garbage data).

**Comment by jwnimmer-tri** [comment]
Oops.  I think my problem was that I was reading the above in too much of a hurry, and I went down the wrong path to trying to repro.  I'll try again from your simpler example.  (I actually tried something just like it, but gave up too quickly.)

**Comment by jwnimmer-tri** [comment]
Actually, I _was_ doing it right.  It turns out that crossing a Jupyter cell boundary is a necessary condition -- I haven't been able to convince the vanilla Python garbage collector to trigger the problem.  However, as you say, it's very easy to reproduce in a notebook.  My fix is proposed in #16095.
